[#27] 회원가입 기능

hyozkim opened the pull request:
- 기능
    - controller에서 JoinRequest를 받은 후 service, repository를 거쳐 데이터베이스 저장
   - JWT 토큰 생성

- 테스트
    - 회원가입 후 user 정상적으로 만들어지는지 데이터 확인
    - 회원가입 시 이메일 중복조회 테스트
    - jwt 토큰 암호화 & 복호화 테스트

hyozkim committed:
update application.yml
- server port 설정
- h2 설정
- database 설정

hyozkim committed:
회원가입 기능 개발

hyozkim committed:
Spring Security 추가
- PasswordEncoder 사용하여 회원 패스워드 암호화
- Security Configure 작성

hyozkim committed:
update gradle dependency
- h2database
- jpa
- spring security
- google guava, lang3

hyozkim committed:
update gradle dependency
- jwt

hyozkim committed:
JWT class 추가
- JWT 생성(헤더,클레임, 시그니처)
- JWT 생성 테스트

hyozkim committed:
테스트 코드 작성
- JWT test
- Service test

hyozkim committed:
update gradle
- 테스트 시 'compileOnly' 종속은 적용 안됨(compileOnly -> complie 수정)

hyozkim committed:
JWT 복호화 기능 추가

hyozkim committed:
로그인 기능 부분 삭제

BaeJi77 commented on back-end/src/main/java/com/bangstagram/user/configure/WebSecurityConfigure.java:
```diff
@@ -0,0 +1,64 @@
+    String issuer;
+    @Value("${jwt.token.secret}")
+    String secret;
+    @Value("${jwt.token.expirySeconds}")
+    int expirySeconds;
```
무제한이면 조금 위험한 거 아니예요?

BaeJi77 reviewed (commented):
jwt 유효시간 무제한이 약간 걸리네요 ㅠㅠ

ike5923 commented on back-end/src/main/java/com/bangstagram/user/controller/UserRestController.java:
```diff
@@ -0,0 +1,43 @@
+
+        String apiToken = user.newJwtToken(jwt,new String[] {"USER_ROLE"});
+
+        return new JoinResult(user, apiToken);
```
return userService.join(joinRequest);

BaeJi77 commented on back-end/src/main/java/com/bangstagram/user/domain/model/user/User.java:
```diff
@@ -0,0 +1,82 @@
+        return jwt.newToken(claims);
+    }
+
+    private static boolean checkAddress(String address) {
+        return matches("[\\w~\\-.+]+@[\\w~\\-]+(\\.[\\w~\\-]+)+", address);
+    }
```
https://javacan.tistory.com/entry/Bean-Validation-2-Spring-5-valiidatiion
@email 참고 하세요~

ike5923 commented on back-end/src/test/java/com/bangstagram/user/service/UserServiceTest.java:
```diff
@@ -0,0 +1,67 @@
+    @Test
+    @Order(2)
+    void 회원가입시_이메일_중복조회한다() {
+        boolean isExist = userService.findByEmail(email).isPresent();
```
exist method로 변경해주세요

hyozkim committed:
코드 리뷰 후 소스 수정
- @Repository 삭제 후 실행 됨..
- Model 부분에 Valid Dto에서 확인하도록 수정
- Service Layer에서 ResponseDto를 생성하여 전달.(controller에서는 전달 역할만)
- boolean check (findByEmail Optional 사용 x) -> jpa existsByEmail 을 사용하여 값 있나 없나 체크
- @Valid 어노테이션 사용
- Dto 끝에 ~Dto 붙이기
- 컨벤션

hyozkim committed:
Merge branch 'develop' into feature/15_signup

hyozkim committed:
update build gradle

hyozkim committed:
update application.yml

BaeJi77 reviewed (approved):
좋아요!! 👏

hyozkim committed:
로그인 기능 개발전으로 되돌리기
- 이슈별 브랜치를 생성해야 하는데 실수로 다른 이슈를 조금 개발을 해버림

hyozkim committed:
Controller Test 코드 작성

BaeJi77 commented on back-end/src/test/java/com/bangstagram/user/controller/UserRestControllerTest.java:
```diff
@@ -0,0 +1,49 @@
+        map.put("email","sa01747@naver.com");
+
+        ObjectMapper mapper = new ObjectMapper();
+        mockMvc.perform(post("http://localhost:9090/users/exists").contentType(MediaType.APPLICATION_JSON).content(mapper.writeValueAsString(map)))
```
여기 http://localhost:9090 이부분은 안쓰시는 게 나을 것 같아요.

hyozkim committed:
review complete!
- Test code 수정

hyozkim committed:
update .gitignore

hyozkim committed:
Merge branch 'develop' into feature/15_signup
